File: Gdx/Historic/fell/core/src/com/dugsolutions/fell/map/gen2/GenZMap.java
package com.dugsolutions.fell.map.gen;

import java.util.ArrayList;
import java.util.Random;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.math.Vector3;
import com.dugsolutions.fell.map.MapGrid;

/**
 * Given a MapGrid, apply a host of z generators across it.
 * 
 * @author dug
 * 
 */
public class GenZMap {

	static final boolean LOG = true;
	static final String TAG = "GenZMap";

	class MyCalcResult implements IGenZ {
		@Override
		public void addHeight(int x, int y, float z) {
			if (LOG) {
				Gdx.app.log(TAG, "addHeight(" + x + ", " + y + ", " + z + ")");
			}
			map.addElevation(x, y, z);
		}

		@Override
		public void addNormal(int x, int y, Vector3 n) {
		}

		@Override
		public boolean genNormal() {
			return false;
		}
	}

	MapGrid map;
	ArrayList<GenZBase> generators = new ArrayList<GenZBase>();
	Random random = new Random();
	int sizeLo;
	int sizeHi;
	MyCalcResult calc = new MyCalcResult();

	public GenZMap(MapGrid m) {
		map = m;
		sizeHi = (map.getMaxIndexX() < map.getMaxIndexY() ? map.getMaxIndexX()
				: map.getMaxIndexY());
		if (sizeHi > 2) {
			sizeLo = 2;
		} else {
			sizeLo = 1;
		}
	}

	public void addGenerator(GenZBase gen) {
		generators.add(gen);
	}

	public void setRandomSeed(long s) {
		random = new Random(s);
	}

	/**
	 * Set the range the bounding box for each gen can be. The X and Y sizes
	 * will randomly be between the indicated range.
	 * 
	 * @param lo
	 *            : lo size in indices
	 * @param hi
	 *            : hi size in indices
	 */
	public void setGenSize(int lo, int hi) {
		sizeLo = lo;
		sizeHi = hi;
	}

	/**
	 * Go through each generator the indicated number of times at random
	 * locations and sizes on the map.
	 * 
	 * @param count
	 */
	public void run() {
		GenZBase gen;
		// Subdivision indices
		final int maxXs = map.getMaxIndexX();
		final int maxYs = map.getMaxIndexY();
		final int sizeLen = sizeHi - sizeLo + 1;
		int xsStart; // subdivision index x
		int ysStart; // subdivision index y
		int xsSize; // subdivision size x
		int ysSize; // subdivision size y;

		while ((gen = getNextGen()) != null) {
			xsSize = random.nextInt(sizeLen) + sizeLo;
			ysSize = random.nextInt(sizeLen) + sizeLo;
			xsStart = random.nextInt(maxXs + 1 - xsSize);
			ysStart = random.nextInt(maxYs + 1 - ysSize);

			gen.gen(calc, xsStart, ysStart, xsStart + xsSize - 1, ysStart
					+ ysSize - 1);

			if (gen.getCount() <= 0) {
				generators.remove(gen);
			}
		}
		map.setElevations();
	}

	GenZBase getNextGen() {
		GenZBase gen;
		if (generators.size() == 0) {
			return null;
		}
		int idx = random.nextInt(generators.size());
		for (int i = 0; i < generators.size(); i++) {
			gen = generators.get((idx + i) % generators.size());
			if (gen.getCount() > 0) {
				return gen;
			}
		}
		return null;
	}
}

File: Gdx/FellDungeon/app/src/main/java/com/dugsolutions/fell/map/gen2/GenZMap.kt
package com.dugsolutions.fell.map.gen2

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.math.MathUtils.random
import com.badlogic.gdx.math.Vector3
import com.dugsolutions.fell.map.MapGrid
import java.util.*

/**
 * Given a MapGrid, apply a host of z generators across it.
 *
 * @author dug
 */
class GenZMap(var map: MapGrid) {

    companion object {
        const val LOG = true
        const val TAG = "GenZMap"
    }

    private var generators = ArrayList<GenZBase>()
    private var random = Random()
    private var sizeLo = 0
    private var sizeHi: Int
    private var calc = MyCalcResult()

    init {
        sizeHi = if (map.maxIndexX < map.maxIndexY) map.maxIndexX else map.maxIndexY
        sizeLo = if (sizeHi > 2) {
            2
        } else {
            1
        }
    }

    inner class MyCalcResult : IGenZ {
        override fun addHeight(x: Int, y: Int, z: Float) {
            if (LOG) {
                Gdx.app.log(TAG, "addHeight($x, $y, $z)")
            }
            map.addElevation(x, y, z)
        }

        override fun addNormal(x: Int, y: Int, n: Vector3) {}
        override fun genNormal(): Boolean {
            return false
        }
    }


    fun addGenerator(gen: GenZBase) {
        generators.add(gen)
    }

    fun setRandomSeed(s: Long) {
        random = Random(s)
    }

    /**
     * Set the range the bounding box for each gen can be. The X and Y sizes
     * will randomly be between the indicated range.
     *
     * @param lo
     * : lo size in indices
     * @param hi
     * : hi size in indices
     */
    fun setGenSize(lo: Int, hi: Int) {
        sizeLo = lo
        sizeHi = hi
    }

    /**
     * Go through each generator the indicated number of times at random
     * locations and sizes on the map.
     */
    fun run() {
        var gen: GenZBase
        // Subdivision indices
        val maxXs = map.maxIndexX
        val maxYs = map.maxIndexY
        val sizeLen = sizeHi - sizeLo + 1
        var xsStart: Int // subdivision index x
        var ysStart: Int // subdivision index y
        var xsSize: Int // subdivision size x
        var ysSize: Int // subdivision size y;
        while (nextGen.also { gen = it!! } != null) {
            xsSize = random.nextInt(sizeLen) + sizeLo
            ysSize = random.nextInt(sizeLen) + sizeLo
            xsStart = random.nextInt(maxXs + 1 - xsSize)
            ysStart = random.nextInt(maxYs + 1 - ysSize)
            gen.gen(
                calc, xsStart, ysStart, xsStart + xsSize - 1, ysStart
                        + ysSize - 1
            )
            if (gen.count <= 0) {
                generators.remove(gen)
            }
        }
        map.setElevations()
    }

    private val nextGen: GenZBase?
        get() {
            var gen: GenZBase
            if (generators.size == 0) {
                return null
            }
            val idx = random.nextInt(generators.size)
            for (i in generators.indices) {
                gen = generators[(idx + i) % generators.size]
                if (gen.count > 0) {
                    return gen
                }
            }
            return null
        }

}